### ISSUE TITLE
Implement pathwise derivative for dirichlet distribution

### ISSUE THREAD

**Issue by neerajprad** [pyro-ppl/numpyro#48]
Refer to #44. We can implement dirichlet sample and grad methods that are similar to the ones we have in PyTorch.

**Comment by fehiepsi** [comment]
More context about this issue:

Dirichlet sampler is available in #81. However, the derivative is based on the gamma sampler which generates dirichlet samples. It is not a correct pathwise derivative. The algorithm for Dirichlet pathwise derivate can be found in [this paper](https://arxiv.org/abs/1806.01851).

**Comment by fehiepsi** [comment]
The current Dirichlet sampler has a reparameterized gradient but its variance would be higher than the version in PyTorch. Given that we mostly rely on the upstream jax samplers now, it is better to use the current high variance version and raise the request to jax team (if needed) instead.

**Comment by adamgayoso** [comment]
Hello,

I'm trying to sample from a Dirichlet with 4 dimensions, but over a batch of thousands of elements. The gradients are very slow compared to PyTorch. Has there been any progress on speeding this up ?

**Comment by fehiepsi** [comment]
Could you try TFP Dirichlet distribution? IIRC it has faster dirichlet sampler.

**Comment by adamgayoso** [comment]
I'm getting nan's there related to dirichlet sampling where I'm not with jax.random.dirichlet/numpyro!...

**Comment by adamgayoso** [comment]
I can confirm the sampling if faster in TFP, but I'll get NaN's during training, which I haven't narrowed down to a gradient issue or a sampling issue.

**Comment by fehiepsi** [comment]
You might want to create a wrapper to [clip Dirichlet samples](https://github.com/pyro-ppl/numpyro/blob/f81ce325de402d9a9891497136a36e3b44f982cc/numpyro/distributions/continuous.py#L261). Something like
```
class RobustDirichlet(...):
    def sample(self, ...):
        samples = super().sample(...)
        return clip(samples)
```

**Comment by martinjankowiak** [comment]
@adamgayoso it might help to make sure your parameter is bounded away from zero e.g. `concentration=0.01 + positive_param`

**Comment by adamgayoso** [comment]
Thank you both for the suggestions, although I'm still getting nan's. I'd need more time to inspect if it's the gradients, but do you know if I can stitch the jax random dirichlet gradient fn with tfp gamma sampler?

**Comment by martinjankowiak** [comment]
```
from tensorflow_probability.substrates.jax import distributions as tfd

def model():
    numpyro.sample("x", tfd.Normal(0, 1))
```

**Comment by adamgayoso** [comment]
Thank you. Where is that happening?

I see this wrapper: https://num.pyro.ai/en/stable/distributions.html#numpyro.contrib.tfp.distributions.TFPDistribution

But not where a custom gradient fn is being set.

**Comment by martinjankowiak** [comment]
i guess to do that you'd to override `sample` and define a custom `jvp` or `jvp` rule

**Comment by adamgayoso** [comment]
@fehiepsi would you be able to check out https://github.com/google/jax/pull/12943 ?

The sampling wasn't the bottleneck, it was actually the lax map for the grads on cpu. I also fixed the lax cond in the sampling loop
